### PR TITLE
Update TableSchema & system tables to resemble ABI V9

### PR DESCRIPTION
**Pull request by kazimuth**
# Description of Changes

This gets rid of most of the usages of `RawModuleDefV8` in `core`. It replaces it with `ModuleDef`, which means that the new validation path is used consistently throughout the codebase. It also refactors the system tables & `TableSchema` types to match the new `ModuleDef`, the `update` functionality to use the new auto-migration functionality based on `ModuleDef`, and does a little bit of other cleanup.

The major user-visible changes from this PR are:
- Some modules that used to be allowed may be rejected by the new, intensive validation rules.
- The `describe` endpoint now returns a json-serialized `RawModuleDefV9` rather than `RawModuleDefV8`.

The major logical changes are:
- `ConstraintSchema` and `IndexSchema` (and their corresponding system tables) no longer store redundant information. Constraints are completely separate from indices; to determine if an index is unique, look up a constraint with its columns. (This is not *quite* true in `mut_tx.rs` and the `table` crate, see my discussion with tyler below, but this is now true throughout more of the codebase.)
- Also note that `IndexSchema` stores a `ColList`, to keep track of field ordering, whereas `ConstraintSchema` stores a `ColSet`, which discards ordering, but can be cheaply constructed from a `ColList`.
- The `Constraints` type is now constructed on-demand from `TableSchema`, rather than being stored directly in `TableSchema`. This means we have a single source of truth for the data in `Constraints` and are less likely to muck it up. `Constraints` is still used in the `vm` and `sql` modules.

# API and ABI breaking changes

ABI break in the system tables, API break at the describe endpoint.

# Expected complexity level and risk

4, this is a system table update and touches things all over `core`.

# Testing

Updated all tests and added some. See also https://github.com/clockworklabs/SpacetimeDB/pull/1737

### CODE REVIEW

File: crates/core/src/db/datastore/system_tables.rs
Comment (kazimuth) [inline crates/core/src/db/datastore/system_tables.rs:260]: Note: I am going to change these to use `add_type`, since `St(...)Row` now implements `SpacetimeType`. I'll add asserts in this method that the fields are in the correct position and have the right labels.

File: crates/core/src/db/datastore/system_tables.rs
Comment (kazimuth) [inline crates/core/src/db/datastore/system_tables.rs:1347]: This is now how all system table rows are going to be read. I haven't finished updating this file yet.

File: crates/primitives/src/ids.rs
Comment (kazimuth) [inline crates/primitives/src/ids.rs:53]: These now need to implement Serialize, Deserialize, and SpacetimeType. They are going to pretend to be their wrapped types.

File: crates/core/src/sql/compiler.rs
Comment (mamcx) [inline crates/core/src/sql/compiler.rs:140]: This will be gone when the new `sql parser` lands, and it will do the proper checkings.

File: crates/lib/src/relation.rs
Comment (mamcx) [inline crates/lib/src/relation.rs:106]:
```suggestion
       proposed_constraints: impl IntoIterator<Item = (ColList, Constraints)>,
```

I think `compression` conveys a different thing.

File: crates/schema/src/schema.rs
Comment (mamcx) [inline crates/schema/src/schema.rs:68]: I think this should instead be a method that returns the `pk` from the `constraints`. There is a good reason to cache this?

Comment (kazimuth) [reply crates/schema/src/schema.rs:68]: There no longer are any `Constraints` stored in the TableSchema. This is the single source of truth for the primary key.

File: crates/core/src/db/datastore/system_tables.rs
Comment (mamcx) [inline crates/core/src/db/datastore/system_tables.rs:287]: A variant `with_primary_key_auto` should exist that also calls the other steps, and is less verbose and error-prone...

File: crates/core/src/db/datastore/system_tables.rs
Comment (mamcx) [inline crates/core/src/db/datastore/system_tables.rs:324]: They used to be comments like `// TODO: Unique constraint on index name?` that this pr made as it was resolved as `no`. Is this the case? If not, then we need to decide about what to do with names.

File: crates/core/src/db/datastore/system_tables.rs
Comment (mamcx) [inline crates/core/src/db/datastore/system_tables.rs:319]:
```suggestion
            "accessor_name_doesnt_matter",
            DEFAULT_ACCESOR_NAME
```

File: crates/core/src/db/update.rs
Comment (kazimuth) [inline crates/core/src/db/update.rs:50]: TODO: need to check for a reducer marked `LifeCycle::Update` and call it here? Or should it be called after auto-migration?

File: crates/core/src/db/datastore/system_tables.rs
Comment (kazimuth) [reply crates/core/src/db/datastore/system_tables.rs:324]: Ah, I'll just put them back in. We should decide that before 1.0 though.

File: crates/lib/src/relation.rs
Comment (kazimuth) [reply crates/lib/src/relation.rs]: I reworked the structure here a little and changed this to `uncombined_constraints`. See the doc comment on `relation.rs:combine_constraints`.

File: crates/core/src/db/datastore/traits.rs
Comment (cloutiertyler) [inline crates/core/src/db/datastore/traits.rs:426]: Why is `is_unique` necessary here? Also should it be included in the `index_schema` or no?

Comment (kazimuth) [reply crates/core/src/db/datastore/traits.rs:426]: See the comment added in `mut_tx.rs`. (It's a little hard to document these APIs because there's multiple layers with nearly-identical structures, not always clear where to put the comments.)

It is still necessary because the `table` crate still takes an `is_unique` parameter. We can eventually move the enforcement of unique constraints outside `table` and then it won't be necessary anymore.

It would be better to just look up the corresponding unique constraint in the system tables, but this might be being called in a transaction where the relevant unique constraint will be created later (in particular during `create_table`), so we need to hedge our bets and give the caller control.

**Comment by kazimuth** [comment]
There's now a fix for the internal tests in private, so this is all ready.

File: crates/core/src/db/update.rs
Comment (kazimuth) [reply crates/core/src/db/update.rs:50]: We're dropping it (https://github.com/clockworklabs/SpacetimeDB/issues/1370), so this is fine.

File: crates/core/src/db/update.rs
Comment (kazimuth) [inline crates/core/src/db/update.rs:4]: This file was completely rewritten so should probably not be reviewed in diff view.

File: crates/core/src/vm.rs
Comment (kazimuth) [inline crates/core/src/vm.rs:544]: This code was dead and now out-of-date so I buried it.

File: crates/core/src/db/datastore/locking_tx_datastore/datastore.rs
Comment (cloutiertyler) [inline crates/core/src/db/datastore/locking_tx_datastore/datastore.rs:1246]: NIT: Replace with `i128::MAX`

File: crates/core/src/db/datastore/system_tables.rs
Comment (cloutiertyler) [inline crates/core/src/db/datastore/system_tables.rs:352]: NIT: This diagram is missing `module_version`.

**Review by cloutiertyler — APPROVED** [review]
I scanned the whole PR. I left a few small nits. AFAICT, this looks really good. Very nice job here @kazimuth.

I am approving this PR from a `traits.rs` perspective and no major objection to the rest, but would probably be good to get another approval as well.

**Review by kim — APPROVED** [review]
Looks good.

Error propagation needs another pass.

File: crates/core/src/db/relational_db.rs
Comment (kazimuth) [reply crates/core/src/db/relational_db.rs:65]: This is a versioning field for the client module, I believe Lightfox requested this. Adding it now for backwards-compatibility.

File: crates/core/src/db/datastore/locking_tx_datastore/datastore.rs
Comment (kazimuth) [reply crates/core/src/db/datastore/locking_tx_datastore/datastore.rs:1246]: I was just copying from the other cases.

File: crates/core/src/host/host_controller.rs
Comment (kazimuth) [reply crates/core/src/host/host_controller.rs:390]: Ah, I refactored `UpdateDatabaseResult` slightly but didn't need to do this here, will rewrite.

File: crates/core/src/db/relational_db.rs
Comment (kim) [reply crates/core/src/db/relational_db.rs:65]: Can we have a docstring telling us that? :cat_with_heart_eyes:

File: crates/core/src/db/datastore/system_tables.rs
Comment (kazimuth) [reply crates/core/src/db/datastore/system_tables.rs:352]: Fixed, and I moved these to doc comments on the `St...Row` types since they make sense there.

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (kazimuth) [reply crates/core/src/host/wasm_common/module_host_actor.rs:343]: Would be a good idea to just log everything that goes through the system loggers yeah. I'll make that a separate issue tho

**Pull request merged**
